otelib/backends/python/base.py

```diff
@@ -0,0 +1,151 @@
+    """Abstract class for strategies for the python backend.
+
+    Parameters:
+        url (str): The base URL of the OTEAPI Service.
+
+    Attributes:
+        url (str): The base URL of the OTEAPI Service.
+        settings (otelib.settings.Settings): OTEAPI Service settings.
+        input_pipe (Optional[Pipe]): An input pipeline.
+
+    """
+
+    # pylint: disable=too-many-instance-attributes
+    # pylint: disable=duplicate-code
+    def __init__(
+        self,
+        url: "Optional[str]" = None,
+    ) -> None:
+        """Initiates a strategy."""
+        if not url:
+            raise ValueError("Url (python) must be specified.")
+
+        self.url: "Optional[str]" = url
+        self.settings: Settings = Settings()
+        self.input_pipe: "Optional[Pipe]" = None
+        self.id: "Optional[str]" = None  # pylint: disable=invalid-name
+
+        # Maybe there is a smarter way of doing abstract attributes
+        self.strategy_name: str = ""
+        self.strategy_config: GenericConfig = GenericConfig
+
+        # Maybe there is a smarter way to have a persistant cache...
+        self.cache = Cache()
+
+        # For debugging/testing
+        self.debug: bool = bool(os.getenv("OTELIB_DEBUG", ""))
+        self._session_id: "Optional[str]" = None
+
+    def create(self, **kwargs) -> None:
+        session_id = kwargs.pop("session_id", None)
+        data = self.strategy_config(**kwargs)
+
+        resource_id = self.strategy_name + "-" + str(uuid4())
+        self.id = resource_id
+        self.cache[resource_id] = data.json()
+
+        if session_id:
+            session = self.cache[session_id]
+            list_key = self.strategy_name + "_info"
+            if list_key in session:
+                session[list_key].extend([resource_id])
+            else:
+                session[list_key] = [resource_id]
+
+    def fetch(self, session_id: str) -> bytes:
+        resource_id = self.id
+
+        config = self.strategy_config(**json.loads(self.cache[resource_id]))
+        session_data = None if not session_id else self.cache[session_id]
+        session_update = create_strategy(self.strategy_name, config)
+        session_update = session_update.get(session=session_data)
+
+        if session_update and session_id:
+            self.cache[session_id].update(session_update)
+
+        return AttrDict(**session_update).json()
+
+    def initialize(self, session_id: str) -> bytes:
+        resource_id = self.id
+
+        config = self.strategy_config(**json.loads(self.cache[resource_id]))
+        if session_id:
+            session_data = self.cache[session_id]
+        else:
+            session_data = None
+
+        strategy = create_strategy(self.strategy_name, config)
+        session_update = strategy.initialize(session=session_data)
+        if session_update and session_id:
+            self.cache[session_id].update(session_update)
+
+        return AttrDict(**session_update).json()
```

Comment:
The .json() method of attribute actually returns a str, maybe this should be changed? I have updated the method signatures to say str instead of bytes, but maybe we should change this at a deeper level, i.e., have .json() of AttrDict return bytes instead

Reply:
> I think this solution is good enough for now. But maybe we can create an issue to make the .json() method of AttrDict return bytes instead of a string?

Let's open an issue and discuss it there. I don't think this is the right thing to do though. But we need to come up with a solution.

Reply:
See https://github.com/EMMC-ASBL/otelib/issues/73